{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MERGE DNDS WITH TARGETS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "targets_df =  pd.read_csv(\"../results/targets_from_xml_selected_none.csv\", index_col=0)\n",
    "dnds_df = pd.read_table(\"../data/ensembl/dn_ds_human_mouse_chimp.csv\")\n",
    "filenames_df = pd.read_csv(\"../data/ensembl/filenames_mapping.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As many targets have non-unique UniProt IDs I decided to remove all ambiquous targets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in total: 593 targets\n",
      "without drugs: 0 targets\n",
      "without UniProt ID: 2 targets\n",
      "\n",
      "# IDs occurences - # of IDs (before filtering):\n",
      "1    571\n",
      "2      8\n",
      "3      2\n",
      "Name: UniProt_ID, dtype: int64\n",
      "\n",
      "# IDs occurences - # of IDs (after filtering):\n",
      "1    571\n",
      "Name: UniProt_ID, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BE0009155</th>\n",
       "      <td>Sn1-specific diacylglycerol lipase alpha</td>\n",
       "      <td>Humans</td>\n",
       "      <td>Y10601</td>\n",
       "      <td>F5GY58</td>\n",
       "      <td>DB14050</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0002442</th>\n",
       "      <td>Toll-like receptor 4</td>\n",
       "      <td>Humans</td>\n",
       "      <td>U93091</td>\n",
       "      <td>O00206</td>\n",
       "      <td>DB05475</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0002356</th>\n",
       "      <td>Voltage-dependent L-type calcium channel subun...</td>\n",
       "      <td>Humans</td>\n",
       "      <td>U95020</td>\n",
       "      <td>O00305</td>\n",
       "      <td>DB01388</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0002406</th>\n",
       "      <td>Phosphatidylinositol 4,5-bisphosphate 3-kinase...</td>\n",
       "      <td>Humans</td>\n",
       "      <td>Y10055</td>\n",
       "      <td>O00329</td>\n",
       "      <td>DB05552</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0002085</th>\n",
       "      <td>cGMP-dependent 3',5'-cyclic phosphodiesterase</td>\n",
       "      <td>Humans</td>\n",
       "      <td>U67733</td>\n",
       "      <td>O00408</td>\n",
       "      <td>DB06246</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        name organism  \\\n",
       "BE0009155           Sn1-specific diacylglycerol lipase alpha   Humans   \n",
       "BE0002442                               Toll-like receptor 4   Humans   \n",
       "BE0002356  Voltage-dependent L-type calcium channel subun...   Humans   \n",
       "BE0002406  Phosphatidylinositol 4,5-bisphosphate 3-kinase...   Humans   \n",
       "BE0002085      cGMP-dependent 3',5'-cyclic phosphodiesterase   Humans   \n",
       "\n",
       "          GenBank_ID UniProt_ID    drugs  drugs_num  \n",
       "BE0009155     Y10601     F5GY58  DB14050          1  \n",
       "BE0002442     U93091     O00206  DB05475          1  \n",
       "BE0002356     U95020     O00305  DB01388          1  \n",
       "BE0002406     Y10055     O00329  DB05552          1  \n",
       "BE0002085     U67733     O00408  DB06246          1  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"in total:\", targets_df.shape[0], \"targets\")\n",
    "print(\"without drugs:\", sum(pd.isnull(targets_df['drugs_num'])), \"targets\")\n",
    "print(\"without UniProt ID:\", sum(targets_df['UniProt_ID'] == '-'), \"targets\")\n",
    "print()\n",
    "print(\"# IDs occurences - # of IDs (before filtering):\")\n",
    "print(targets_df['UniProt_ID'].value_counts().value_counts())\n",
    "\n",
    "targets_df = targets_df[targets_df['UniProt_ID'] != '-']  # remove targets without UniProt IDs\n",
    "targets_df.drop_duplicates(subset =\"UniProt_ID\", keep = False, inplace = True)  # remove targets with non-unique UniProt IDs\n",
    "\n",
    "print()\n",
    "print(\"# IDs occurences - # of IDs (after filtering):\")\n",
    "print(targets_df['UniProt_ID'].value_counts().value_counts())\n",
    "\n",
    "targets_df.sort_values('UniProt_ID').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at `filenames_df`. Some gene IDs are mapped to multiple UniProt IDs. I decided to leave it that way so we won't lose UniProt IDs when merging with `targets_df`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(129449, 2)\n",
      "\n",
      "# IDs occurences - # of IDs:\n",
      "1     62479\n",
      "2      3323\n",
      "7       477\n",
      "8       437\n",
      "3       300\n",
      "6       193\n",
      "5       189\n",
      "4       154\n",
      "10       40\n",
      "31       31\n",
      "42       28\n",
      "36       24\n",
      "14       17\n",
      "41       17\n",
      "33        9\n",
      "28        5\n",
      "9         4\n",
      "39        3\n",
      "40        3\n",
      "13        3\n",
      "12        2\n",
      "15        2\n",
      "35        2\n",
      "30        1\n",
      "Name: UniProt_ID, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID UniProt_ID\n",
       "0  ENSG00000210049        NaN\n",
       "1  ENSG00000211459        NaN\n",
       "2  ENSG00000210077        NaN\n",
       "3  ENSG00000210082        NaN\n",
       "4  ENSG00000209082        NaN"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(filenames_df.shape)\n",
    "print()\n",
    "\n",
    "filenames_df.columns = ['Gene stable ID', 'UniProt_ID'] # rename the column so it can be used for merging\n",
    "\n",
    "print(\"# IDs occurences - # of IDs:\")\n",
    "print(filenames_df['UniProt_ID'].value_counts().value_counts())\n",
    "\n",
    "filenames_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge `targets_df` with `filenames_df` first so every target can be then mapped to its gene in `dnds_df`. One should not merge `filenames_df` with `dnds_df` first as `filenames_df` includes multiple UniProt IDs for each gene. Merging with `targets_df` leaves only needed IDs present in `targets_df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(631, 7)\n",
      "19 targets unmapped\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "      <th>Gene stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C-1-tetrahydrofolate synthase, cytoplasmic</td>\n",
       "      <td>Humans</td>\n",
       "      <td>J04031</td>\n",
       "      <td>P11586</td>\n",
       "      <td>DB00116</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bifunctional methylenetetrahydrofolate dehydro...</td>\n",
       "      <td>Humans</td>\n",
       "      <td>X16396</td>\n",
       "      <td>P13995</td>\n",
       "      <td>DB00116</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000065911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aminomethyltransferase, mitochondrial</td>\n",
       "      <td>Humans</td>\n",
       "      <td>D13811</td>\n",
       "      <td>P48728</td>\n",
       "      <td>DB00116; DB04789</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000145020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cytosolic 10-formyltetrahydrofolate dehydrogenase</td>\n",
       "      <td>Humans</td>\n",
       "      <td>AF052732</td>\n",
       "      <td>O75891</td>\n",
       "      <td>DB00116</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000144908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Methionine synthase</td>\n",
       "      <td>Humans</td>\n",
       "      <td>U71285</td>\n",
       "      <td>Q99707</td>\n",
       "      <td>DB00116</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000116984</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name organism GenBank_ID  \\\n",
       "0         C-1-tetrahydrofolate synthase, cytoplasmic   Humans     J04031   \n",
       "1  Bifunctional methylenetetrahydrofolate dehydro...   Humans     X16396   \n",
       "2              Aminomethyltransferase, mitochondrial   Humans     D13811   \n",
       "3  Cytosolic 10-formyltetrahydrofolate dehydrogenase   Humans   AF052732   \n",
       "4                                Methionine synthase   Humans     U71285   \n",
       "\n",
       "  UniProt_ID             drugs  drugs_num   Gene stable ID  \n",
       "0     P11586           DB00116          1              NaN  \n",
       "1     P13995           DB00116          1  ENSG00000065911  \n",
       "2     P48728  DB00116; DB04789          2  ENSG00000145020  \n",
       "3     O75891           DB00116          1  ENSG00000144908  \n",
       "4     Q99707           DB00116          1  ENSG00000116984  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets_mapped_df = pd.merge(targets_df, filenames_df, how='left', on='UniProt_ID')\n",
    "print(targets_mapped_df.shape)\n",
    "print(sum(pd.isnull(targets_mapped_df['Gene stable ID'])), \"targets unmapped\")\n",
    "targets_mapped_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now look at `dnds_df`. How many Human genes do have both orthologs and calculated dN, dS?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "with ortholog (for Chimpanzee): 25484 genes\n",
      "with dN/dS (for Chimpanzee): 25057 genes\n",
      "with ortholog (for Mouse): 29290 genes\n",
      "with dN/dS (for Mouse): 25649 genes\n",
      "with both orthologs: 20522 genes\n",
      "with both dN/dS: 22603 genes\n",
      "in total: 77255 genes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>mitochondrially encoded tRNA-Phe (UUU/C) [Sour...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>mitochondrially encoded 12S rRNA [Source:HGNC ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>mitochondrially encoded tRNA-Val (GUN) [Source...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>mitochondrially encoded 16S rRNA [Source:HGNC ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000210049     MT-TF   \n",
       "1  ENSG00000211459   MT-RNR1   \n",
       "2  ENSG00000210077     MT-TV   \n",
       "3  ENSG00000210082   MT-RNR2   \n",
       "4  ENSG00000209082    MT-TL1   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded tRNA-Phe (UUU/C) [Sour...                  NaN   \n",
       "1  mitochondrially encoded 12S rRNA [Source:HGNC ...                  NaN   \n",
       "2  mitochondrially encoded tRNA-Val (GUN) [Source...                  NaN   \n",
       "3  mitochondrially encoded 16S rRNA [Source:HGNC ...                  NaN   \n",
       "4  mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...                  NaN   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0                 NaN                 NaN             NaN            NaN   \n",
       "1                 NaN                 NaN             NaN            NaN   \n",
       "2                 NaN                 NaN             NaN            NaN   \n",
       "3                 NaN                 NaN             NaN            NaN   \n",
       "4                 NaN                 NaN             NaN            NaN   \n",
       "\n",
       "   dS with Mouse  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"with ortholog (for Chimpanzee):\", sum(pd.notnull(dnds_df['Chimpanzee gene name'])), \"genes\")\n",
    "print(\"with dN/dS (for Chimpanzee):\", sum(pd.notnull(dnds_df['dN with Chimpanzee'])), \"genes\")\n",
    "print(\"with ortholog (for Mouse):\", sum(pd.notnull(dnds_df['Mouse gene name'])), \"genes\")\n",
    "print(\"with dN/dS (for Mouse):\", sum(pd.notnull(dnds_df['dN with Mouse'])), \"genes\")\n",
    "print(\"with both orthologs:\", sum(pd.notnull(dnds_df['Mouse gene name']) & pd.notnull(dnds_df['Chimpanzee gene name'])), \"genes\")\n",
    "print(\"with both dN/dS:\", sum(pd.notnull(dnds_df['dN with Mouse']) & pd.notnull(dnds_df['dN with Chimpanzee'])), \"genes\")\n",
    "\n",
    "print(\"in total:\", dnds_df.shape[0], \"genes\")\n",
    "\n",
    "dnds_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many genes have 1:1 orthologs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of human genes - # of orthologs\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1      64458\n",
       "2       1810\n",
       "3        300\n",
       "4        166\n",
       "5         88\n",
       "6         52\n",
       "7         33\n",
       "14        26\n",
       "9         22\n",
       "8         20\n",
       "93        13\n",
       "58        12\n",
       "10        10\n",
       "186        9\n",
       "15         8\n",
       "12         6\n",
       "110        6\n",
       "29         6\n",
       "19         5\n",
       "28         5\n",
       "17         4\n",
       "13         4\n",
       "11         3\n",
       "16         3\n",
       "116        3\n",
       "27         3\n",
       "42         2\n",
       "20         2\n",
       "25         2\n",
       "54         1\n",
       "21         1\n",
       "26         1\n",
       "23         1\n",
       "22         1\n",
       "18         1\n",
       "Name: Gene stable ID, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"# of human genes - # of orthologs\")\n",
    "dnds_df['Gene stable ID'].value_counts(dropna=False).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Leave only genes with 1:1 orthologs present both in Chimpanzee and Mouse:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before filtering: 77255 genes\n",
      "\n",
      "with orthologs in both organisms: 22603 genes\n",
      "\n",
      "with 1:1 orthologs only: 14903 genes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>mitochondrially encoded ATP synthase membrane ...</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.3331</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>0.4871</td>\n",
       "      <td>0.8480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.6559</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>0.3455</td>\n",
       "      <td>13.3498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000169926</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>Kruppel like factor 13 [Source:HGNC Symbol;Acc...</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>Klf13</td>\n",
       "      <td>0.0219</td>\n",
       "      <td>0.6810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000276289</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>potassium voltage-gated channel subfamily E re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.0649</td>\n",
       "      <td>Kcne1</td>\n",
       "      <td>0.1382</td>\n",
       "      <td>0.8135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000067992</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>pyruvate dehydrogenase kinase 3 [Source:HGNC S...</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>Pdk3</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.5213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000228253   MT-ATP8   \n",
       "1  ENSG00000198695    MT-ND6   \n",
       "2  ENSG00000169926     KLF13   \n",
       "3  ENSG00000276289    KCNE1B   \n",
       "4  ENSG00000067992      PDK3   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded ATP synthase membrane ...              MT-ATP8   \n",
       "1  mitochondrially encoded NADH:ubiquinone oxidor...               MT-ND6   \n",
       "2  Kruppel like factor 13 [Source:HGNC Symbol;Acc...                KLF13   \n",
       "3  potassium voltage-gated channel subfamily E re...                  NaN   \n",
       "4  pyruvate dehydrogenase kinase 3 [Source:HGNC S...                 PDK3   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0              0.0325              0.3331         mt-Atp8         0.4871   \n",
       "1              0.0204              0.6559          mt-Nd6         0.3455   \n",
       "2              0.0026              0.0771           Klf13         0.0219   \n",
       "3              0.0165              0.0649           Kcne1         0.1382   \n",
       "4              0.0152              0.0300            Pdk3         0.0151   \n",
       "\n",
       "   dS with Mouse  \n",
       "0         0.8480  \n",
       "1        13.3498  \n",
       "2         0.6810  \n",
       "3         0.8135  \n",
       "4         0.5213  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('before filtering:', dnds_df.shape[0], 'genes')\n",
    "print()\n",
    "\n",
    "dnds_sel_df = dnds_df.dropna(subset=['dN with Mouse','dN with Chimpanzee','dS with Mouse','dS with Chimpanzee'])\n",
    "print('with orthologs in both organisms:', dnds_sel_df.shape[0], 'genes')\n",
    "print()\n",
    "\n",
    "dnds_sel_df = dnds_sel_df.drop_duplicates(subset = [\"Gene stable ID\"], keep = False).reset_index(drop=True)\n",
    "print('with 1:1 orthologs only:', dnds_sel_df.shape[0], 'genes')\n",
    "\n",
    "dnds_sel_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create `grand_df` which is `dnds_sel_df` merged with `targets_mapped_df`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14904, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>mitochondrially encoded ATP synthase membrane ...</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.3331</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>0.4871</td>\n",
       "      <td>0.8480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.6559</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>0.3455</td>\n",
       "      <td>13.3498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000169926</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>Kruppel like factor 13 [Source:HGNC Symbol;Acc...</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>Klf13</td>\n",
       "      <td>0.0219</td>\n",
       "      <td>0.6810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000276289</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>potassium voltage-gated channel subfamily E re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.0649</td>\n",
       "      <td>Kcne1</td>\n",
       "      <td>0.1382</td>\n",
       "      <td>0.8135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000067992</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>pyruvate dehydrogenase kinase 3 [Source:HGNC S...</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>Pdk3</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.5213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000228253   MT-ATP8   \n",
       "1  ENSG00000198695    MT-ND6   \n",
       "2  ENSG00000169926     KLF13   \n",
       "3  ENSG00000276289    KCNE1B   \n",
       "4  ENSG00000067992      PDK3   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded ATP synthase membrane ...              MT-ATP8   \n",
       "1  mitochondrially encoded NADH:ubiquinone oxidor...               MT-ND6   \n",
       "2  Kruppel like factor 13 [Source:HGNC Symbol;Acc...                KLF13   \n",
       "3  potassium voltage-gated channel subfamily E re...                  NaN   \n",
       "4  pyruvate dehydrogenase kinase 3 [Source:HGNC S...                 PDK3   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0              0.0325              0.3331         mt-Atp8         0.4871   \n",
       "1              0.0204              0.6559          mt-Nd6         0.3455   \n",
       "2              0.0026              0.0771           Klf13         0.0219   \n",
       "3              0.0165              0.0649           Kcne1         0.1382   \n",
       "4              0.0152              0.0300            Pdk3         0.0151   \n",
       "\n",
       "   dS with Mouse name organism GenBank_ID UniProt_ID drugs  drugs_num  \n",
       "0         0.8480  NaN      NaN        NaN        NaN   NaN        NaN  \n",
       "1        13.3498  NaN      NaN        NaN        NaN   NaN        NaN  \n",
       "2         0.6810  NaN      NaN        NaN        NaN   NaN        NaN  \n",
       "3         0.8135  NaN      NaN        NaN        NaN   NaN        NaN  \n",
       "4         0.5213  NaN      NaN        NaN        NaN   NaN        NaN  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grand_df = pd.merge(dnds_sel_df, targets_mapped_df, how='left', on='Gene stable ID')\n",
    "print(grand_df.shape)\n",
    "grand_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add columns with dN/dS ratios and whether the gene is present in `targets_mapped_df` or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "      <th>name</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "      <th>dN/dS with Chimpanzee</th>\n",
       "      <th>dN/dS with Mouse</th>\n",
       "      <th>present in DrugBank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>mitochondrially encoded ATP synthase membrane ...</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.3331</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>0.4871</td>\n",
       "      <td>0.8480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0976</td>\n",
       "      <td>0.5744</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.6559</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>0.3455</td>\n",
       "      <td>13.3498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0311</td>\n",
       "      <td>0.0259</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000169926</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>Kruppel like factor 13 [Source:HGNC Symbol;Acc...</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>Klf13</td>\n",
       "      <td>0.0219</td>\n",
       "      <td>0.6810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0337</td>\n",
       "      <td>0.0322</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000276289</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>potassium voltage-gated channel subfamily E re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.0649</td>\n",
       "      <td>Kcne1</td>\n",
       "      <td>0.1382</td>\n",
       "      <td>0.8135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2542</td>\n",
       "      <td>0.1699</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000067992</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>pyruvate dehydrogenase kinase 3 [Source:HGNC S...</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>Pdk3</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.5213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5067</td>\n",
       "      <td>0.0290</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000228253   MT-ATP8   \n",
       "1  ENSG00000198695    MT-ND6   \n",
       "2  ENSG00000169926     KLF13   \n",
       "3  ENSG00000276289    KCNE1B   \n",
       "4  ENSG00000067992      PDK3   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded ATP synthase membrane ...              MT-ATP8   \n",
       "1  mitochondrially encoded NADH:ubiquinone oxidor...               MT-ND6   \n",
       "2  Kruppel like factor 13 [Source:HGNC Symbol;Acc...                KLF13   \n",
       "3  potassium voltage-gated channel subfamily E re...                  NaN   \n",
       "4  pyruvate dehydrogenase kinase 3 [Source:HGNC S...                 PDK3   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0              0.0325              0.3331         mt-Atp8         0.4871   \n",
       "1              0.0204              0.6559          mt-Nd6         0.3455   \n",
       "2              0.0026              0.0771           Klf13         0.0219   \n",
       "3              0.0165              0.0649           Kcne1         0.1382   \n",
       "4              0.0152              0.0300            Pdk3         0.0151   \n",
       "\n",
       "   dS with Mouse name GenBank_ID UniProt_ID drugs  drugs_num  \\\n",
       "0         0.8480  NaN        NaN        NaN   NaN        NaN   \n",
       "1        13.3498  NaN        NaN        NaN   NaN        NaN   \n",
       "2         0.6810  NaN        NaN        NaN   NaN        NaN   \n",
       "3         0.8135  NaN        NaN        NaN   NaN        NaN   \n",
       "4         0.5213  NaN        NaN        NaN   NaN        NaN   \n",
       "\n",
       "   dN/dS with Chimpanzee  dN/dS with Mouse  present in DrugBank  \n",
       "0                 0.0976            0.5744                False  \n",
       "1                 0.0311            0.0259                False  \n",
       "2                 0.0337            0.0322                False  \n",
       "3                 0.2542            0.1699                False  \n",
       "4                 0.5067            0.0290                False  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grand_df['dN/dS with Chimpanzee'] = round(grand_df['dN with Chimpanzee'] / grand_df['dS with Chimpanzee'], 4)\n",
    "grand_df['dN/dS with Mouse'] = round(grand_df['dN with Mouse'] / grand_df['dS with Mouse'], 4)\n",
    "grand_df['present in DrugBank'] = pd.notna(grand_df['drugs_num'])\n",
    "grand_df.drop(columns=\"organism\", inplace=True)\n",
    "grand_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some genes have both dN and dS equal to 0. For these genes dN/dS is set to 'NaN':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2970 genes with dN/dS = NaN (Chimpanzee)\n",
      "256 genes with dN/dS = NaN (Mouse)\n"
     ]
    }
   ],
   "source": [
    "print(sum(pd.isna(grand_df['dN/dS with Chimpanzee'])), 'genes with dN/dS = NaN (Chimpanzee)')\n",
    "#print(sum((grand_df['dN with Chimpanzee'] == 0) & (grand_df['dS with Chimpanzee'] == 0)))\n",
    "\n",
    "print(sum(pd.isna(grand_df['dN/dS with Mouse'])), 'genes with dN/dS = NaN (Mouse)')\n",
    "#print(sum((grand_df['dN with Mouse'] == 0) & (grand_df['dS with Mouse'] == 0)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export `grand_df` to a `.csv` file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "grand_df[['Gene name', 'Gene description', 'dN with Chimpanzee', 'dS with Chimpanzee', 'dN/dS with Chimpanzee', 'dN with Mouse', 'dS with Mouse', 'dN/dS with Mouse', 'present in DrugBank']].to_csv('../results/dnds_for_targets_none.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
