{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MERGE DNDS WITH TARGETS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "targets_df =  pd.read_csv(\"../results/targets_from_xml_selected.csv\", index_col=0)\n",
    "dnds_df = pd.read_table(\"../data/ensembl/dn_ds_human_mouse_chimp.csv\")\n",
    "filenames_df = pd.read_csv(\"../data/ensembl/filenames_mapping.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As many targets have non-unique UniProt IDs I decided to remove all ambiquous targets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in total: 2555 targets\n",
      "without drugs: 0 targets\n",
      "without UniProt ID: 53 targets\n",
      "\n",
      "# IDs occurences - # of IDs (before filtering):\n",
      "1     2437\n",
      "2       22\n",
      "3        7\n",
      "53       1\n",
      "Name: UniProt_ID, dtype: int64\n",
      "\n",
      "# IDs occurences - # of IDs (after filtering):\n",
      "1    2437\n",
      "Name: UniProt_ID, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BE0003619</th>\n",
       "      <td>Gamma-aminobutyric acid receptor subunit rho-3</td>\n",
       "      <td>Humans</td>\n",
       "      <td>AC026100</td>\n",
       "      <td>A8MPY1</td>\n",
       "      <td>DB00231; DB00546; DB00690; DB00801; DB00829; D...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0003600</th>\n",
       "      <td>BCR/ABL fusion protein isoform X9</td>\n",
       "      <td>Humans</td>\n",
       "      <td>EU216066</td>\n",
       "      <td>A9UF02</td>\n",
       "      <td>DB00619</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0009155</th>\n",
       "      <td>Sn1-specific diacylglycerol lipase alpha</td>\n",
       "      <td>Humans</td>\n",
       "      <td>Y10601</td>\n",
       "      <td>F5GY58</td>\n",
       "      <td>DB14050</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0004105</th>\n",
       "      <td>Serine/threonine-protein kinase Sgk1</td>\n",
       "      <td>Humans</td>\n",
       "      <td>Y10032</td>\n",
       "      <td>O00141</td>\n",
       "      <td>DB03247; DB07837; DB08191</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BE0004639</th>\n",
       "      <td>Thymidine kinase 2, mitochondrial</td>\n",
       "      <td>Humans</td>\n",
       "      <td>X56793</td>\n",
       "      <td>O00142</td>\n",
       "      <td>DB02452; DB02594; DB04485</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     name organism GenBank_ID  \\\n",
       "BE0003619  Gamma-aminobutyric acid receptor subunit rho-3   Humans   AC026100   \n",
       "BE0003600               BCR/ABL fusion protein isoform X9   Humans   EU216066   \n",
       "BE0009155        Sn1-specific diacylglycerol lipase alpha   Humans     Y10601   \n",
       "BE0004105            Serine/threonine-protein kinase Sgk1   Humans     Y10032   \n",
       "BE0004639               Thymidine kinase 2, mitochondrial   Humans     X56793   \n",
       "\n",
       "          UniProt_ID                                              drugs  \\\n",
       "BE0003619     A8MPY1  DB00231; DB00546; DB00690; DB00801; DB00829; D...   \n",
       "BE0003600     A9UF02                                            DB00619   \n",
       "BE0009155     F5GY58                                            DB14050   \n",
       "BE0004105     O00141                          DB03247; DB07837; DB08191   \n",
       "BE0004639     O00142                          DB02452; DB02594; DB04485   \n",
       "\n",
       "           drugs_num  \n",
       "BE0003619         15  \n",
       "BE0003600          1  \n",
       "BE0009155          1  \n",
       "BE0004105          3  \n",
       "BE0004639          3  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"in total:\", targets_df.shape[0], \"targets\")\n",
    "print(\"without drugs:\", sum(pd.isnull(targets_df['drugs_num'])), \"targets\")\n",
    "print(\"without UniProt ID:\", sum(targets_df['UniProt_ID'] == '-'), \"targets\")\n",
    "print()\n",
    "print(\"# IDs occurences - # of IDs (before filtering):\")\n",
    "print(targets_df['UniProt_ID'].value_counts().value_counts())\n",
    "\n",
    "targets_df = targets_df[targets_df['UniProt_ID'] != '-']  # remove targets without UniProt IDs\n",
    "targets_df.drop_duplicates(subset =\"UniProt_ID\", keep = False, inplace = True)  # remove targets with non-unique UniProt IDs\n",
    "\n",
    "print()\n",
    "print(\"# IDs occurences - # of IDs (after filtering):\")\n",
    "print(targets_df['UniProt_ID'].value_counts().value_counts())\n",
    "\n",
    "targets_df.sort_values('UniProt_ID').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at `filenames_df`. Some gene IDs are mapped to multiple UniProt IDs. I decided to leave it that way so we won't lose UniProt IDs when merging with `targets_df`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(129449, 2)\n",
      "\n",
      "# IDs occurences - # of IDs:\n",
      "1     62479\n",
      "2      3323\n",
      "7       477\n",
      "8       437\n",
      "3       300\n",
      "6       193\n",
      "5       189\n",
      "4       154\n",
      "10       40\n",
      "31       31\n",
      "42       28\n",
      "36       24\n",
      "14       17\n",
      "41       17\n",
      "33        9\n",
      "28        5\n",
      "9         4\n",
      "39        3\n",
      "40        3\n",
      "13        3\n",
      "12        2\n",
      "15        2\n",
      "35        2\n",
      "30        1\n",
      "Name: UniProt_ID, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID UniProt_ID\n",
       "0  ENSG00000210049        NaN\n",
       "1  ENSG00000211459        NaN\n",
       "2  ENSG00000210077        NaN\n",
       "3  ENSG00000210082        NaN\n",
       "4  ENSG00000209082        NaN"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(filenames_df.shape)\n",
    "print()\n",
    "\n",
    "filenames_df.columns = ['Gene stable ID', 'UniProt_ID'] # rename the column so it can be used for merging\n",
    "\n",
    "print(\"# IDs occurences - # of IDs:\")\n",
    "print(filenames_df['UniProt_ID'].value_counts().value_counts())\n",
    "\n",
    "filenames_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge `targets_df` with `filenames_df` first so every target can be then mapped to its gene in `dnds_df`. One should not merge `filenames_df` with `dnds_df` first as `filenames_df` includes multiple UniProt IDs for each gene. Merging with `targets_df` leaves only needed IDs present in `targets_df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2698, 7)\n",
      "92 targets unmapped\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "      <th>Gene stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alanine--glyoxylate aminotransferase 2, mitoch...</td>\n",
       "      <td>Humans</td>\n",
       "      <td>AJ292204</td>\n",
       "      <td>Q9BYV1</td>\n",
       "      <td>DB00114</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000113492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glutamate decarboxylase 1</td>\n",
       "      <td>Humans</td>\n",
       "      <td>M81883</td>\n",
       "      <td>Q99259</td>\n",
       "      <td>DB00114; DB00142</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000128683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cystathionine beta-synthase</td>\n",
       "      <td>Humans</td>\n",
       "      <td>L19501</td>\n",
       "      <td>P35520</td>\n",
       "      <td>DB00114; DB00118; DB00133; DB00151</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000160200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Kynureninase</td>\n",
       "      <td>Humans</td>\n",
       "      <td>U57721</td>\n",
       "      <td>Q16719</td>\n",
       "      <td>DB00114; DB07069</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000115919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Serine hydroxymethyltransferase, cytosolic</td>\n",
       "      <td>Humans</td>\n",
       "      <td>L11931</td>\n",
       "      <td>P34896</td>\n",
       "      <td>DB00114; DB00116; DB01055; DB02067; DB02800; D...</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000176974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name organism GenBank_ID  \\\n",
       "0  Alanine--glyoxylate aminotransferase 2, mitoch...   Humans   AJ292204   \n",
       "1                          Glutamate decarboxylase 1   Humans     M81883   \n",
       "2                        Cystathionine beta-synthase   Humans     L19501   \n",
       "3                                       Kynureninase   Humans     U57721   \n",
       "4         Serine hydroxymethyltransferase, cytosolic   Humans     L11931   \n",
       "\n",
       "  UniProt_ID                                              drugs  drugs_num  \\\n",
       "0     Q9BYV1                                            DB00114          1   \n",
       "1     Q99259                                   DB00114; DB00142          2   \n",
       "2     P35520                 DB00114; DB00118; DB00133; DB00151          4   \n",
       "3     Q16719                                   DB00114; DB07069          2   \n",
       "4     P34896  DB00114; DB00116; DB01055; DB02067; DB02800; D...          6   \n",
       "\n",
       "    Gene stable ID  \n",
       "0  ENSG00000113492  \n",
       "1  ENSG00000128683  \n",
       "2  ENSG00000160200  \n",
       "3  ENSG00000115919  \n",
       "4  ENSG00000176974  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets_mapped_df = pd.merge(targets_df, filenames_df, how='left', on='UniProt_ID')\n",
    "print(targets_mapped_df.shape)\n",
    "print(sum(pd.isnull(targets_mapped_df['Gene stable ID'])), \"targets unmapped\")\n",
    "targets_mapped_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now look at `dnds_df`. How many Human genes do have both orthologs and calculated dN, dS?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "with ortholog (for Chimpanzee): 25484 genes\n",
      "with dN/dS (for Chimpanzee): 25057 genes\n",
      "with ortholog (for Mouse): 29290 genes\n",
      "with dN/dS (for Mouse): 25649 genes\n",
      "with both orthologs: 20522 genes\n",
      "with both dN/dS: 22603 genes\n",
      "in total: 77255 genes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>mitochondrially encoded tRNA-Phe (UUU/C) [Sour...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>mitochondrially encoded 12S rRNA [Source:HGNC ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>mitochondrially encoded tRNA-Val (GUN) [Source...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>mitochondrially encoded 16S rRNA [Source:HGNC ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000210049     MT-TF   \n",
       "1  ENSG00000211459   MT-RNR1   \n",
       "2  ENSG00000210077     MT-TV   \n",
       "3  ENSG00000210082   MT-RNR2   \n",
       "4  ENSG00000209082    MT-TL1   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded tRNA-Phe (UUU/C) [Sour...                  NaN   \n",
       "1  mitochondrially encoded 12S rRNA [Source:HGNC ...                  NaN   \n",
       "2  mitochondrially encoded tRNA-Val (GUN) [Source...                  NaN   \n",
       "3  mitochondrially encoded 16S rRNA [Source:HGNC ...                  NaN   \n",
       "4  mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...                  NaN   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0                 NaN                 NaN             NaN            NaN   \n",
       "1                 NaN                 NaN             NaN            NaN   \n",
       "2                 NaN                 NaN             NaN            NaN   \n",
       "3                 NaN                 NaN             NaN            NaN   \n",
       "4                 NaN                 NaN             NaN            NaN   \n",
       "\n",
       "   dS with Mouse  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"with ortholog (for Chimpanzee):\", sum(pd.notnull(dnds_df['Chimpanzee gene name'])), \"genes\")\n",
    "print(\"with dN/dS (for Chimpanzee):\", sum(pd.notnull(dnds_df['dN with Chimpanzee'])), \"genes\")\n",
    "print(\"with ortholog (for Mouse):\", sum(pd.notnull(dnds_df['Mouse gene name'])), \"genes\")\n",
    "print(\"with dN/dS (for Mouse):\", sum(pd.notnull(dnds_df['dN with Mouse'])), \"genes\")\n",
    "print(\"with both orthologs:\", sum(pd.notnull(dnds_df['Mouse gene name']) & pd.notnull(dnds_df['Chimpanzee gene name'])), \"genes\")\n",
    "print(\"with both dN/dS:\", sum(pd.notnull(dnds_df['dN with Mouse']) & pd.notnull(dnds_df['dN with Chimpanzee'])), \"genes\")\n",
    "\n",
    "print(\"in total:\", dnds_df.shape[0], \"genes\")\n",
    "\n",
    "dnds_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many genes have 1:1 orthologs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of human genes - # of orthologs\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1      64458\n",
       "2       1810\n",
       "3        300\n",
       "4        166\n",
       "5         88\n",
       "6         52\n",
       "7         33\n",
       "14        26\n",
       "9         22\n",
       "8         20\n",
       "93        13\n",
       "58        12\n",
       "10        10\n",
       "186        9\n",
       "15         8\n",
       "12         6\n",
       "110        6\n",
       "29         6\n",
       "19         5\n",
       "28         5\n",
       "17         4\n",
       "13         4\n",
       "11         3\n",
       "16         3\n",
       "116        3\n",
       "27         3\n",
       "42         2\n",
       "20         2\n",
       "25         2\n",
       "54         1\n",
       "21         1\n",
       "26         1\n",
       "23         1\n",
       "22         1\n",
       "18         1\n",
       "Name: Gene stable ID, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"# of human genes - # of orthologs\")\n",
    "dnds_df['Gene stable ID'].value_counts(dropna=False).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Leave only genes with 1:1 orthologs present both in Chimpanzee and Mouse:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before filtering: 77255 genes\n",
      "\n",
      "with orthologs in both organisms: 22603 genes\n",
      "\n",
      "with 1:1 orthologs only: 14903 genes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>mitochondrially encoded ATP synthase membrane ...</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.3331</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>0.4871</td>\n",
       "      <td>0.8480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.6559</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>0.3455</td>\n",
       "      <td>13.3498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000169926</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>Kruppel like factor 13 [Source:HGNC Symbol;Acc...</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>Klf13</td>\n",
       "      <td>0.0219</td>\n",
       "      <td>0.6810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000276289</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>potassium voltage-gated channel subfamily E re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.0649</td>\n",
       "      <td>Kcne1</td>\n",
       "      <td>0.1382</td>\n",
       "      <td>0.8135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000067992</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>pyruvate dehydrogenase kinase 3 [Source:HGNC S...</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>Pdk3</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.5213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000228253   MT-ATP8   \n",
       "1  ENSG00000198695    MT-ND6   \n",
       "2  ENSG00000169926     KLF13   \n",
       "3  ENSG00000276289    KCNE1B   \n",
       "4  ENSG00000067992      PDK3   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded ATP synthase membrane ...              MT-ATP8   \n",
       "1  mitochondrially encoded NADH:ubiquinone oxidor...               MT-ND6   \n",
       "2  Kruppel like factor 13 [Source:HGNC Symbol;Acc...                KLF13   \n",
       "3  potassium voltage-gated channel subfamily E re...                  NaN   \n",
       "4  pyruvate dehydrogenase kinase 3 [Source:HGNC S...                 PDK3   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0              0.0325              0.3331         mt-Atp8         0.4871   \n",
       "1              0.0204              0.6559          mt-Nd6         0.3455   \n",
       "2              0.0026              0.0771           Klf13         0.0219   \n",
       "3              0.0165              0.0649           Kcne1         0.1382   \n",
       "4              0.0152              0.0300            Pdk3         0.0151   \n",
       "\n",
       "   dS with Mouse  \n",
       "0         0.8480  \n",
       "1        13.3498  \n",
       "2         0.6810  \n",
       "3         0.8135  \n",
       "4         0.5213  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('before filtering:', dnds_df.shape[0], 'genes')\n",
    "print()\n",
    "\n",
    "dnds_sel_df = dnds_df.dropna(subset=['dN with Mouse','dN with Chimpanzee','dS with Mouse','dS with Chimpanzee'])\n",
    "print('with orthologs in both organisms:', dnds_sel_df.shape[0], 'genes')\n",
    "print()\n",
    "\n",
    "dnds_sel_df = dnds_sel_df.drop_duplicates(subset = [\"Gene stable ID\"], keep = False).reset_index(drop=True)\n",
    "print('with 1:1 orthologs only:', dnds_sel_df.shape[0], 'genes')\n",
    "\n",
    "dnds_sel_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create `grand_df` which is `dnds_sel_df` merged with `targets_mapped_df`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14907, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>mitochondrially encoded ATP synthase membrane ...</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.3331</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>0.4871</td>\n",
       "      <td>0.8480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.6559</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>0.3455</td>\n",
       "      <td>13.3498</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6</td>\n",
       "      <td>Humans</td>\n",
       "      <td>J01415</td>\n",
       "      <td>P03923</td>\n",
       "      <td>DB00157</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000169926</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>Kruppel like factor 13 [Source:HGNC Symbol;Acc...</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>Klf13</td>\n",
       "      <td>0.0219</td>\n",
       "      <td>0.6810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000276289</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>potassium voltage-gated channel subfamily E re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.0649</td>\n",
       "      <td>Kcne1</td>\n",
       "      <td>0.1382</td>\n",
       "      <td>0.8135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000067992</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>pyruvate dehydrogenase kinase 3 [Source:HGNC S...</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>Pdk3</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.5213</td>\n",
       "      <td>[Pyruvate dehydrogenase [lipoamide]] kinase is...</td>\n",
       "      <td>Humans</td>\n",
       "      <td>L42452</td>\n",
       "      <td>Q15120</td>\n",
       "      <td>DB03758; DB03760</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000228253   MT-ATP8   \n",
       "1  ENSG00000198695    MT-ND6   \n",
       "2  ENSG00000169926     KLF13   \n",
       "3  ENSG00000276289    KCNE1B   \n",
       "4  ENSG00000067992      PDK3   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded ATP synthase membrane ...              MT-ATP8   \n",
       "1  mitochondrially encoded NADH:ubiquinone oxidor...               MT-ND6   \n",
       "2  Kruppel like factor 13 [Source:HGNC Symbol;Acc...                KLF13   \n",
       "3  potassium voltage-gated channel subfamily E re...                  NaN   \n",
       "4  pyruvate dehydrogenase kinase 3 [Source:HGNC S...                 PDK3   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0              0.0325              0.3331         mt-Atp8         0.4871   \n",
       "1              0.0204              0.6559          mt-Nd6         0.3455   \n",
       "2              0.0026              0.0771           Klf13         0.0219   \n",
       "3              0.0165              0.0649           Kcne1         0.1382   \n",
       "4              0.0152              0.0300            Pdk3         0.0151   \n",
       "\n",
       "   dS with Mouse                                               name organism  \\\n",
       "0         0.8480                                                NaN      NaN   \n",
       "1        13.3498             NADH-ubiquinone oxidoreductase chain 6   Humans   \n",
       "2         0.6810                                                NaN      NaN   \n",
       "3         0.8135                                                NaN      NaN   \n",
       "4         0.5213  [Pyruvate dehydrogenase [lipoamide]] kinase is...   Humans   \n",
       "\n",
       "  GenBank_ID UniProt_ID             drugs  drugs_num  \n",
       "0        NaN        NaN               NaN        NaN  \n",
       "1     J01415     P03923           DB00157        1.0  \n",
       "2        NaN        NaN               NaN        NaN  \n",
       "3        NaN        NaN               NaN        NaN  \n",
       "4     L42452     Q15120  DB03758; DB03760        2.0  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grand_df = pd.merge(dnds_sel_df, targets_mapped_df, how='left', on='Gene stable ID')\n",
    "print(grand_df.shape)\n",
    "grand_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add columns with dN/dS ratios and whether the gene is present in `targets_mapped_df` or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chimpanzee gene name</th>\n",
       "      <th>dN with Chimpanzee</th>\n",
       "      <th>dS with Chimpanzee</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>dN with Mouse</th>\n",
       "      <th>dS with Mouse</th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>GenBank_ID</th>\n",
       "      <th>UniProt_ID</th>\n",
       "      <th>drugs</th>\n",
       "      <th>drugs_num</th>\n",
       "      <th>dN/dS with Chimpanzee</th>\n",
       "      <th>dN/dS with Mouse</th>\n",
       "      <th>present in DrugBank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>mitochondrially encoded ATP synthase membrane ...</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>0.0325</td>\n",
       "      <td>0.3331</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>0.4871</td>\n",
       "      <td>0.8480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0976</td>\n",
       "      <td>0.5744</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.6559</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>0.3455</td>\n",
       "      <td>13.3498</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6</td>\n",
       "      <td>Humans</td>\n",
       "      <td>J01415</td>\n",
       "      <td>P03923</td>\n",
       "      <td>DB00157</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0311</td>\n",
       "      <td>0.0259</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000169926</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>Kruppel like factor 13 [Source:HGNC Symbol;Acc...</td>\n",
       "      <td>KLF13</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>Klf13</td>\n",
       "      <td>0.0219</td>\n",
       "      <td>0.6810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0337</td>\n",
       "      <td>0.0322</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000276289</td>\n",
       "      <td>KCNE1B</td>\n",
       "      <td>potassium voltage-gated channel subfamily E re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0165</td>\n",
       "      <td>0.0649</td>\n",
       "      <td>Kcne1</td>\n",
       "      <td>0.1382</td>\n",
       "      <td>0.8135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2542</td>\n",
       "      <td>0.1699</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000067992</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>pyruvate dehydrogenase kinase 3 [Source:HGNC S...</td>\n",
       "      <td>PDK3</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>Pdk3</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0.5213</td>\n",
       "      <td>[Pyruvate dehydrogenase [lipoamide]] kinase is...</td>\n",
       "      <td>Humans</td>\n",
       "      <td>L42452</td>\n",
       "      <td>Q15120</td>\n",
       "      <td>DB03758; DB03760</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.5067</td>\n",
       "      <td>0.0290</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene name  \\\n",
       "0  ENSG00000228253   MT-ATP8   \n",
       "1  ENSG00000198695    MT-ND6   \n",
       "2  ENSG00000169926     KLF13   \n",
       "3  ENSG00000276289    KCNE1B   \n",
       "4  ENSG00000067992      PDK3   \n",
       "\n",
       "                                    Gene description Chimpanzee gene name  \\\n",
       "0  mitochondrially encoded ATP synthase membrane ...              MT-ATP8   \n",
       "1  mitochondrially encoded NADH:ubiquinone oxidor...               MT-ND6   \n",
       "2  Kruppel like factor 13 [Source:HGNC Symbol;Acc...                KLF13   \n",
       "3  potassium voltage-gated channel subfamily E re...                  NaN   \n",
       "4  pyruvate dehydrogenase kinase 3 [Source:HGNC S...                 PDK3   \n",
       "\n",
       "   dN with Chimpanzee  dS with Chimpanzee Mouse gene name  dN with Mouse  \\\n",
       "0              0.0325              0.3331         mt-Atp8         0.4871   \n",
       "1              0.0204              0.6559          mt-Nd6         0.3455   \n",
       "2              0.0026              0.0771           Klf13         0.0219   \n",
       "3              0.0165              0.0649           Kcne1         0.1382   \n",
       "4              0.0152              0.0300            Pdk3         0.0151   \n",
       "\n",
       "   dS with Mouse                                               name organism  \\\n",
       "0         0.8480                                                NaN      NaN   \n",
       "1        13.3498             NADH-ubiquinone oxidoreductase chain 6   Humans   \n",
       "2         0.6810                                                NaN      NaN   \n",
       "3         0.8135                                                NaN      NaN   \n",
       "4         0.5213  [Pyruvate dehydrogenase [lipoamide]] kinase is...   Humans   \n",
       "\n",
       "  GenBank_ID UniProt_ID             drugs  drugs_num  dN/dS with Chimpanzee  \\\n",
       "0        NaN        NaN               NaN        NaN                 0.0976   \n",
       "1     J01415     P03923           DB00157        1.0                 0.0311   \n",
       "2        NaN        NaN               NaN        NaN                 0.0337   \n",
       "3        NaN        NaN               NaN        NaN                 0.2542   \n",
       "4     L42452     Q15120  DB03758; DB03760        2.0                 0.5067   \n",
       "\n",
       "   dN/dS with Mouse  present in DrugBank  \n",
       "0            0.5744                False  \n",
       "1            0.0259                 True  \n",
       "2            0.0322                False  \n",
       "3            0.1699                False  \n",
       "4            0.0290                 True  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grand_df['dN/dS with Chimpanzee'] = round(grand_df['dN with Chimpanzee'] / grand_df['dS with Chimpanzee'], 4)\n",
    "grand_df['dN/dS with Mouse'] = round(grand_df['dN with Mouse'] / grand_df['dS with Mouse'], 4)\n",
    "grand_df['present in DrugBank'] = pd.notna(grand_df['drugs_num'])\n",
    "grand_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some genes have both dN and dS equal to 0. For these genes dN/dS is set to 'NaN':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2971 genes with dN/dS = NaN (Chimpanzee)\n",
      "256 genes with dN/dS = NaN (Mouse)\n"
     ]
    }
   ],
   "source": [
    "print(sum(pd.isna(grand_df['dN/dS with Chimpanzee'])), 'genes with dN/dS = NaN (Chimpanzee)')\n",
    "#print(sum((grand_df['dN with Chimpanzee'] == 0) & (grand_df['dS with Chimpanzee'] == 0)))\n",
    "\n",
    "print(sum(pd.isna(grand_df['dN/dS with Mouse'])), 'genes with dN/dS = NaN (Mouse)')\n",
    "#print(sum((grand_df['dN with Mouse'] == 0) & (grand_df['dS with Mouse'] == 0)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export `grand_df` to a `.csv` file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "grand_df[['Gene name', 'Gene description', 'dN/dS with Chimpanzee', 'dN/dS with Mouse', 'present in DrugBank']].to_csv('../results/dnds_for_targets.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
